File: android/src/main/kotlin/dev/pam/scanner/QrImageModule.kt
package dev.pam.scanner

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Bitmap
import android.net.Uri
import android.os.Handler
import android.os.Looper
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import dev.pam.nativeapp.modules.ModuleCompletion
import dev.pam.nativeapp.modules.ModuleResultStatus
import dev.pam.nativeapp.modules.NativeModule
import dev.pam.nativeapp.protocol.WireMap
import dev.pam.nativeapp.protocol.WireValue
import org.json.JSONArray
import java.io.File
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class QrImageModule(context: Context) : NativeModule {
    private val context = context.applicationContext
    private val executor = ThreadPoolExecutor(0, 1, 30, TimeUnit.SECONDS, LinkedBlockingQueue())
    private val main = Handler(Looper.getMainLooper())
    private val busy = AtomicBoolean(false)

    override fun invoke(method: String, payload: ByteArray, completion: ModuleCompletion) {
        if (method != "decodeQrImage" || !busy.compareAndSet(false, true)) {
            fail(completion, "QR image decoder is unavailable or busy.")
            return
        }
        executor.execute {
            try {
                val values = WireMap.decode(payload)
                val source = (values["uri"] as? WireValue.Text)?.value ?: error("Missing image URI")
                require(source.length <= 8192)
                val uri = Uri.parse(source)
                require(uri.scheme == "content" || uri.scheme == "file" || uri.scheme == "pam-file")
                val bitmap = readBitmap(uri)
                val scanner = try {
                    BarcodeScanning.getClient(BarcodeScannerOptions.Builder().setBarcodeFormats(Barcode.FORMAT_QR_CODE).build())
                } catch (error: Exception) {
                    bitmap.recycle()
                    throw error
                }
                try {
                    scanner.process(InputImage.fromBitmap(bitmap, 0))
                        .addOnCompleteListener { task ->
                            val codes = if (task.isSuccessful) task.result.mapNotNull { it.rawValue }.filter { it.isNotEmpty() }.distinct().take(16) else null
                            scanner.close()
                            bitmap.recycle()
                            busy.set(false)
                            if (codes == null) {
                                fail(completion, "Unable to decode the selected image.")
                            } else {
                                completion.complete(ModuleResultStatus.SUCCESS, WireMap.encode(mapOf("values" to WireValue.Text(JSONArray(codes).toString()))))
                            }
                        }
                } catch (error: Exception) {
                    scanner.close()
                    bitmap.recycle()
                    throw error
                }
            } catch (_: Exception) {
                busy.set(false)
                fail(completion, "Unable to decode the selected image.")
            }
        }
    }

    private fun fail(completion: ModuleCompletion, message: String) {
        main.post { completion.complete(ModuleResultStatus.FAILURE, message.toByteArray()) }
    }

    private fun readBitmap(uri: Uri): Bitmap {
        val snapshot = File.createTempFile("pam-qr-", ".image", context.cacheDir)
        try {
            val source = if (uri.scheme == "pam-file") resolvePamFile(context.filesDir, uri) else uri
            context.contentResolver.openInputStream(source).use { input ->
                requireNotNull(input)
                snapshot.outputStream().use { ImageInput.copyBounded(input, it) }
            }
            val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            BitmapFactory.decodeFile(snapshot.path, bounds)
            val options = BitmapFactory.Options().apply {
                inSampleSize = ImageDecodeSize.sample(bounds.outWidth, bounds.outHeight)
            }
            return requireNotNull(BitmapFactory.decodeFile(snapshot.path, options))
        } finally {
            snapshot.delete()
        }
    }
}

internal fun resolvePamFile(filesDirectory: File, uri: Uri): Uri {
    require(uri.scheme.equals("pam-file", ignoreCase = true) && uri.authority.isNullOrEmpty())
    require(uri.query == null && uri.fragment == null)
    val segments = uri.pathSegments
    require(segments.isNotEmpty() && segments.all {
        it.isNotEmpty() && it != "." && it != ".." && '/' !in it && '\\' !in it &&
            it.none { character -> character.code < 0x20 || character.code == 0x7f }
    })
    val root = File(filesDirectory, "pam-files").canonicalFile
    val source = File(root, segments.joinToString(File.separator)).canonicalFile
    require(source.toPath().startsWith(root.toPath()) && source.isFile)
    return Uri.fromFile(source)
}

File: ios/Sources/QrImageModule.swift
import Foundation
import PamNative

public final class QrImageModule: NativeModule, @unchecked Sendable {
    private let queue = DispatchQueue(label: "pam.scanner.image", qos: .userInitiated)
    private let available = DispatchSemaphore(value: 1)

    public init() {}

    public func invoke(method: String, payload: Data, completion: @escaping ModuleCompletion) {
        guard method == "decodeQrImage", available.wait(timeout: .now()) == .success else {
            completion(.failure, Data("QR image decoder is unavailable or busy.".utf8))
            return
        }
        queue.async {
            let result: Result<Data, Error> = Result {
                let values = try WireMap.decode(payload)
                guard case let .text(uri)? = values["uri"], uri.utf8.count <= 8192 else {
                    throw ImageError.invalid
                }
                let url = try Self.sourceURL(uri)
                let codes = try QrImageDecoder.decode(url)
                let json = try JSONSerialization.data(withJSONObject: codes)
                return try WireMap.encode(["values": .text(String(decoding: json, as: UTF8.self))])
            }
            self.available.signal()
            DispatchQueue.main.async {
                switch result {
                case .success(let data): completion(.success, data)
                case .failure: completion(.failure, Data("Unable to decode the selected image.".utf8))
                }
            }
        }
    }

    private static func sourceURL(_ source: String) throws -> URL {
        guard let uri = URLComponents(string: source), uri.query == nil, uri.fragment == nil else {
            throw ImageError.invalid
        }
        if uri.scheme?.lowercased() == "file", let url = uri.url, url.isFileURL { return url }
        guard uri.scheme?.lowercased() == "pam-file", uri.host?.isEmpty ?? true else {
            throw ImageError.invalid
        }
        let path = uri.percentEncodedPath.removingPercentEncoding ?? ""
        let segments = path.split(separator: "/", omittingEmptySubsequences: true).map(String.init)
        guard !segments.isEmpty, segments.allSatisfy({ segment in
            !segment.isEmpty && segment != "." && segment != ".." &&
                !segment.contains("/") && !segment.contains("\\") &&
                !segment.unicodeScalars.contains(where: { $0.value < 0x20 || $0.value == 0x7f })
        }) else { throw ImageError.invalid }
        let root = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("pam-files", isDirectory: true).resolvingSymlinksInPath()
        let url = segments.reduce(root) { $0.appendingPathComponent($1) }.resolvingSymlinksInPath()
        let values = try? url.resourceValues(forKeys: [.isRegularFileKey])
        guard url.path.hasPrefix(root.path + "/"), values?.isRegularFile == true else {
            throw ImageError.invalid
        }
        return url
    }
}

private enum ImageError: Error { case invalid }
